quark.c

```diff
@@ -408,43 +479,179 @@ process_cache_inherit(struct quark_queue *qq, struct quark_process *qp, int ppid
-			break;
-		process_cache_delete(qq, qp);
-		n++;
+	return (0);
+}
+
+/*
+ * Socket stuff
+ */
+
+#if 0
+static int
+sockaddr_to_str(const struct quark_sockaddr *qsa, char *buf, size_t len)
+{
+	size_t	n;
+	int	r;
+
+	if (inet_ntop(qsa->af, qsa->addr6, buf, len) == NULL)
+		return (-1);
+	n = strlen(buf);
+	buf += n;
+	len -= n;
+	if (len == 0)
+		return (-1);
+	r = snprintf(buf, len, ":%d", ntohs(qsa->port));
+	if (r <= 0 || r >= (int)len)
+		return (-1);
+
+	return (0);
+}
+
+static int
+socket_to_str(const struct quark_socket *qsk, char *buf, size_t len)
+{
+	size_t	n;
+
+	if (sockaddr_to_str(&qsk->local, buf, len) == -1)
+		return (-1);
+	n = strlcat(buf, " -> ", len);
+	if (n >= len)
+		return (-1);
+	buf += n;
+	len -= n;
+	if (sockaddr_to_str(&qsk->remote, buf, len) == -1)
+		return (-1);
+
+	return (0);
+}
+#endif
+
+static struct quark_socket *
+socket_cache_get(struct quark_queue *qq,
+    struct quark_sockaddr *local, struct quark_sockaddr *remote, int alloc)
+{
+	struct quark_socket	 key;
+	struct quark_socket	*qsk;
+
+	if ((local->af != AF_INET && local->af != AF_INET6) ||
+	    (remote->af != AF_INET && remote->af != AF_INET6) ||
+	    (local->af != remote->af))
+		return (errno = EINVAL, NULL);
+
+	key.local = *local;
+	key.remote = *remote;
+	qsk = RB_FIND(socket_by_src_dst, &qq->socket_by_src_dst, &key);
+	if (qsk != NULL)
+		return (qsk);
+
+	if (!alloc) {
+		errno = ESRCH;
+		return (NULL);
 	}
 
-	return (n);
+	qsk = calloc(1, sizeof(*qsk));
+	if (qsk == NULL)
+		return (NULL);
+	qsk->local = key.local;
+	qsk->remote = key.remote;
+	if (RB_INSERT(socket_by_src_dst, &qq->socket_by_src_dst, qsk) != NULL) {
+		err(1, "collision, this is a bug");
+		free(qsk);
+		return (NULL);
+	}
+
+	return (qsk);
+}
+
+static void
+socket_cache_delete(struct quark_queue *qq, struct quark_socket *qsk)
+{
+	RB_REMOVE(socket_by_src_dst, &qq->socket_by_src_dst, qsk);
+	gc_unlink(qq, &qsk->gc);
+	free(qsk);
 }
 
 static int
-process_by_pid_cmp(struct quark_process *a, struct quark_process *b)
+socket_by_src_dst_cmp(struct quark_socket *a, struct quark_socket *b)
 {
-	if (a->pid < b->pid)
+	size_t	cmplen;
+	int	r;
+
+	if (a->remote.port < b->remote.port)
 		return (-1);
-	else if (a->pid > b->pid)
+	else if (a->remote.port > b->remote.port)
 		return (1);
 
-	return (0);
+	if (a->local.port < b->local.port)
+		return (-1);
+	else if (a->local.port > b->local.port)
+		return (1);
+
+	if (a->remote.af < b->remote.af)
+		return (-1);
+	else if (a->remote.af > b->remote.af)
+		return (1);
+
+	if (a->local.af < b->local.af)
+		return (-1);
+	else if (a->local.af > b->local.af)
+		return (1);
+
+	cmplen = a->remote.af == AF_INET ? 4 : 16;
+	r = memcmp(&a->remote.addr6, &b->remote.addr6, cmplen);
+	if (r != 0)
+		return (r);
+
+	cmplen = a->local.af == AF_INET ? 4 : 16;
+	r = memcmp(&a->local.addr6, &b->local.addr6, cmplen);
+
+	return (r);
+}
+
+const struct quark_socket *
+quark_socket_lookup(struct quark_queue *qq,
```

Comment:
I don't see this function being called. Enlighten me please.

Reply:
Oh, I suppose these are used by library users.

Reply:
Yep!